starport/services/scaffolder/module.go

```diff
@@ -213,36 +267,40 @@ func moduleExists(appPath string, moduleName string) (bool, error) {
 		return false, nil
 	}
 
-	return true, err
+	return err == nil, err
 }
 
-func checkModuleName(moduleName string) error {
+// checkModuleName checks if the name can be used as a module name
+func checkModuleName(appPath, moduleName string) error {
 	// go keyword
 	if token.Lookup(moduleName).IsKeyword() {
 		return fmt.Errorf("%s is a Go keyword", moduleName)
 	}
 
-	// name of default registered module
-	switch moduleName {
-	case
-		"auth",
-		"genutil",
-		"bank",
-		"capability",
-		"staking",
-		"mint",
-		"distr",
-		"gov",
-		"params",
-		"crisis",
-		"slashing",
-		"ibc",
-		"upgrade",
-		"evidence",
-		"transfer",
-		"vesting":
-		return fmt.Errorf("%s is a default module", moduleName)
+	// check if the name is a reserved name
+	if _, ok := reservedNames[moduleName]; ok {
+		return fmt.Errorf("%s is a reserved name and can't be used as a module name", moduleName)
 	}
+
+	// check if the name can imply potential store key collision
+	for _, defaultStoreKey := range defaultStoreKeys {
+		if strings.HasPrefix(moduleName, defaultStoreKey) {
+			return fmt.Errorf("the module name can't be prefixed with %s because of potential store key collision", defaultStoreKey)
+		}
+	}
+
+	// check store key with user's defined modules
+	// we consider all user's defined modules use the module name as the store key
+	entries, err := os.ReadDir(filepath.Join(appPath, moduleDir))
+	if err != nil {
+		return err
+	}
+	for _, entry := range entries {
+		if strings.HasPrefix(moduleName, entry.Name()) {
+			return fmt.Errorf("the module name can't be prefixed with %s because of potential store key collision", entry.Name())
+		}
+	}
```

Comment:
```suggestion
	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		if strings.HasPrefix(moduleName, entry.Name()) {
			return fmt.Errorf("the module name can't be prefixed with %s because of potential store key collision", entry.Name())
		}
	}
	}
```